File: simulator/coordinator/internal/simulation/testdata/irp/retailer.py
from __future__ import annotations 
 
from typing import Any 
 
from modeling import Atomic, RunnableModelCfg, RunnableModelPortCfg, new_atomic_from_cfg 
 
MINUTES_PER_DAY = 24 * 60 
 
 
class RetailerIRP(Atomic): 
    def __init__(self, id: str, name: str, ports=None): 
        super().__init__(id=id, name=name, ports=ports) 
        self.parameters: dict[str, Any] = {} 
        self.current_time: int = 0 
        self.next_close_at: int = 16 * 60 
        self.current_inventory: float = 0.0 
        self.retailer_id: int = 0 
        self.min_inventory: float = 0.0 
        self.max_inventory: float = 0.0 
        self.daily_consumption: float = 0.0 
        self.inventory_cost: float = 0.0 
 
    def initialize(self) -> None: 
        self.current_time = 0 
        self.retailer_id = _as_int(self.parameters.get("retailer_id"), 0) 
        self.current_inventory = _as_float(self.parameters.get("starting_inventory"), 0.0) 
        self.min_inventory = _as_float(self.parameters.get("min_inventory"), 0.0) 
        self.max_inventory = _as_float(self.parameters.get("max_inventory"), 0.0) 
        self.daily_consumption = _as_float(self.parameters.get("daily_consumption"), 0.0) 
        self.inventory_cost = _as_float(self.parameters.get("inventory_cost"), 0.0) 
        closing_hour = _as_int(self.parameters.get("closing_hour"), 16) 
        self.next_close_at = closing_hour * 60 
 
    def exit(self) -> None: 
        return 
 
    def ta(self) -> float: 
        return max(float(self.next_close_at - self.current_time), 0.0) 
 
    def delt_int(self) -> None: 
        self.current_time = self.next_close_at 
 
        if self.current_inventory > self.max_inventory: 
            print( 
                f"Retailer {self.retailer_id} exceeded max inventory: " 
                f"current={self.current_inventory} max={self.max_inventory}" 
            ) 

        self.current_inventory -= self.daily_consumption 

        if self.current_inventory < self.min_inventory: 
            print( 
                f"Retailer {self.retailer_id} below min inventory: " 
                f"current={self.current_inventory} min={self.min_inventory}" 
            )
 
        self.next_close_at += MINUTES_PER_DAY 
 
    def delt_ext(self, e: float) -> None: 
        self.current_time += int(round(e)) 
 
        try: 
            in_port = self.get_port_by_name("receiveDelivery") 
        except KeyError: 
            return 
 
        for value in list(in_port.get_values()): 
            if not isinstance(value, dict): 
                continue 
            delivery_retailer_id = _as_int(value.get("retailerId"), -1) 
            if delivery_retailer_id != self.retailer_id: 
                continue 
            self.current_inventory += _as_float(value.get("productAmount"), 0.0) 
 
        in_port.clear() 
 
    def delt_con(self, e: float) -> None: 
        self.delt_int() 
        self.delt_ext(0.0) 
 
    def lambda_(self) -> None: 
        predicted_inventory = self.current_inventory - self.daily_consumption 
        payload = { 
            "day": _day_from_minute(self.next_close_at), 
            "retailerId": self.retailer_id, 
            "cost": predicted_inventory * self.inventory_cost, 
        } 

        try: 
            out_port = self.get_port_by_name("dailyInventoryCost") 
        except KeyError: 
            return 
        out_port.add_value(payload)
 
 
def NewModel(config: dict) -> Atomic: 
    raw_ports = config.get("ports") or [] 
    ports_cfg = [ 
        RunnableModelPortCfg( 
            id=p["id"], 
            name=p.get("name", p["id"]), 
            type=p["type"], 
        ) 
        for p in raw_ports 
    ] 

    cfg = RunnableModelCfg( 
        id=config["id"], 
        name=config["name"], 
        ports=ports_cfg, 
    ) 

    model = new_atomic_from_cfg(cfg, RetailerIRP) 
    raw_parameters = config.get("parameters") or [] 
    model.parameters = { 
        p["name"]: p.get("value") 
        for p in raw_parameters 
        if isinstance(p, dict) and p.get("name") 
    } 
    return model
 
 
def _as_int(value: Any, fallback: int = 0) -> int: 
    try: 
        return int(value) 
    except (TypeError, ValueError): 
        return fallback 
 
 
def _as_float(value: Any, fallback: float = 0.0) -> float: 
    try: 
        return float(value) 
    except (TypeError, ValueError): 
        return fallback 
 
 
def _day_from_minute(minute: int) -> int: 
    return minute // MINUTES_PER_DAY + 1 

File: simulator/coordinator/internal/simulation/testdata/irp_multi/retailer.go
package main

import (
	"fmt"
	"math"
	"strconv"

	modeling "devsforge-wrapper/modeling"
)

const minutesPerDay int64 = 24 * 60

type RetailerIRP struct {
	modeling.Atomic
	parameters       map[string]interface{}
	currentTime      int64
	nextCloseAt      int64
	currentInventory float64
	retailerID       int
	minInventory     float64
	maxInventory     float64
	dailyConsumption float64
	inventoryCost    float64
	closingHour      int
}

func NewModel(cfg modeling.RunnableModel) modeling.Atomic {
	params := make(map[string]interface{}, len(cfg.Parameters))
	for _, p := range cfg.Parameters {
		params[p.Name] = p.Value
	}
	return &RetailerIRP{
		Atomic:     modeling.NewAtomic(cfg),
		parameters: params,
	}
}

func (m *RetailerIRP) Initialize() {
	m.currentTime = 0
	m.retailerID = asIntRetailer(m.parameters["retailer_id"], 0)
	m.currentInventory = asFloatRetailer(m.parameters["starting_inventory"], 0)
	m.minInventory = asFloatRetailer(m.parameters["min_inventory"], 0)
	m.maxInventory = asFloatRetailer(m.parameters["max_inventory"], 0)
	m.dailyConsumption = asFloatRetailer(m.parameters["daily_consumption"], 0)
	m.inventoryCost = asFloatRetailer(m.parameters["inventory_cost"], 0)
	m.closingHour = asIntRetailer(m.parameters["closing_hour"], 16)
	m.nextCloseAt = int64(m.closingHour) * 60
}

func (m *RetailerIRP) Exit() {}

func (m *RetailerIRP) TA() float64 {
	delta := float64(m.nextCloseAt - m.currentTime)
	if delta < 0 {
		return 0
	}
	return delta
}

func (m *RetailerIRP) DeltInt() {
	m.currentTime = m.nextCloseAt

	if m.currentInventory > m.maxInventory {
		fmt.Printf("Retailer %d exceeded max inventory: current=%f max=%f\n", m.retailerID, m.currentInventory, m.maxInventory)
	}

	m.currentInventory -= m.dailyConsumption

	if m.currentInventory < m.minInventory {
		fmt.Printf("Retailer %d below min inventory: current=%f min=%f\n", m.retailerID, m.currentInventory, m.minInventory)
	}

	m.nextCloseAt += minutesPerDay
}

func (m *RetailerIRP) DeltExt(e float64) {
	m.currentTime += int64(math.Round(e))

	port, err := m.GetPortByName("receiveDelivery")
	if err != nil {
		return
	}
	if values, ok := port.GetValues().([]interface{}); ok {
		for _, value := range values {
			delivery, ok := asMapRetailer(value)
			if !ok {
				continue
			}
			deliveryRetailerID := asIntRetailer(delivery["retailerId"], -1)
			if deliveryRetailerID != m.retailerID {
				continue
			}
			m.currentInventory += asFloatRetailer(delivery["productAmount"], 0)
		}
	}
	port.Clear()
}

func (m *RetailerIRP) DeltCon(e float64) {
	m.DeltInt()
	m.DeltExt(0)
}

func (m *RetailerIRP) Lambda() {
	predictedInventory := m.currentInventory - m.dailyConsumption
	payload := map[string]interface{}{
		"day":        dayFromMinute(m.nextCloseAt),
		"retailerId": m.retailerID,
		"cost":       predictedInventory * m.inventoryCost,
	}

	port, err := m.GetPortByName("dailyInventoryCost")
	if err != nil {
		return
	}
	port.AddValue(payload)
}

func dayFromMinute(minute int64) int {
	if minute < 0 {
		return 1
	}
	return int(minute/minutesPerDay) + 1
}

func asMapRetailer(value interface{}) (map[string]interface{}, bool) {
	m, ok := value.(map[string]interface{})
	return m, ok
}

func asIntRetailer(value interface{}, fallback int) int {
	switch v := value.(type) {
	case int:
		return v
	case int64:
		return int(v)
	case int32:
		return int(v)
	case float64:
		return int(v)
	case float32:
		return int(v)
	case string:
		if p, err := strconv.Atoi(v); err == nil {
			return p
		}
	}
	return fallback
}

func asFloatRetailer(value interface{}, fallback float64) float64 {
	switch v := value.(type) {
	case float64:
		return v
	case float32:
		return float64(v)
	case int:
		return float64(v)
	case int64:
		return float64(v)
	case int32:
		return float64(v)
	case string:
		if p, err := strconv.ParseFloat(v, 64); err == nil {
			return p
		}
	}
	return fallback
}
